### PR TITLE
Add toml

### PR DESCRIPTION
**Pull request by pie-flavor**

### CODE REVIEW

**Comment by Meronat** [comment]
Can you update the maven for latest version releases and update to the latest toml4j version(0.7.2?) and make sure your implementation still works? 

Since you haven't updated this in a while, I'm assuming there might be other changes you're interested in making. 

Once that happens I will test this more thoroughly myself as well.

**Comment by pie-flavor** [comment]
Nice. Travis failed because Maven is looking for the public version of configurate-core instead of my modified version.

**Comment by Meronat** [comment]
Can you use `Nullable` annotations where appropriate?

**Comment by XakepSDK** [comment]
I disagree support of `java.util.Date`, old time api must die.

**Comment by Meronat** [comment]
@XakepSDK is right. `Instant` should be enough, but if not, there are plenty other options in the `java.time` library such as `LocalDate`, `LocalDateTime`, `OffsetDateTime`, and more.

**Comment by pie-flavor** [comment]
@parlough @XakepSDK I don't know what's wrong with leaving it, since it's just there for conversion reasons. Seems weird to *remove* support for something after already adding it without having maintenance costs.

**Comment by pie-flavor** [comment]
@XakepSDK Good point.

File: configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLOptions.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+import com.moandjiezana.toml.TomlWriter;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import java.time.ZoneOffset;
+import java.util.TimeZone;
+
+/**
+ * Represents various options for TOML saving.
+ */
+public class TOMLOptions {
```
Comment (lucko) [inline configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLOptions.java]: IMO, the options encapsulated by this class should just be expressed on the loader. This is the layout used by all of the other loader impls.

I've done it this way in my implementation here.
https://github.com/lucko/configurate/blob/lucko-master/configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLConfigurationLoader.java#L62-L172

Feel free to copy.

File: configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLConfigurationLoader.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    @Override
+    protected void loadInternal(ConfigurationNode node, BufferedReader reader) throws IOException {
+        Toml toml = new Toml().read(reader);
+        Map<String, Object> map = toml.toMap();
+        fixToml(node, map);
+    }
+
+    @SuppressWarnings("unchecked")
+    void fixToml(@NonNull ConfigurationNode node, @NonNull Map<String, Object> input) {
```
Comment (lucko) [inline configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLConfigurationLoader.java]: `#fixToml` and '#breakToml` aren't very obvious method names, it's not clear what they do.

I'd prefer clearer names, such as read/write, load/save object.

File: configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLConfigurationLoader.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        input.forEach((str, obj) -> {
```
Comment (lucko) [inline configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLConfigurationLoader.java]: Just use a regular loop? The key `String#replace` call on the line below can be made inline.

**Comment by parlough** [comment]
Also I believe this was talked about elsewhere, but putting it here as a reminder. We should probably switch to a better maintained library with support for the latest TOML standard, which looks like https://github.com/ConsenSys/cava/tree/master/toml for the moment.

**Comment by pie-flavor** [comment]
@lucko You're welcome to commit to `feature/toml` directly. I have that setting enabled.

**Comment by pie-flavor** [comment]
@parlough I cloned it and it fails its own unit tests. Toml4j may not be 100% compliant but my wrapper code makes it so you can't tell the difference; the same can't be said of Cava's.

**Comment by XakepSDK** [comment]
toml4j if 0.4.0 ver impl and cava is 0.5.0 ver impl.

**Comment by XakepSDK** [comment]
Implementation links here: https://github.com/toml-lang/toml/wiki

**Comment by parlough** [comment]
^ Exactly. It's not a good idea to invest time in a library that likely won't be updated and is already behind on the latest spec. If there are issues with Cava's we can always PR changes as they seem to be quick enough with those. If we want we can create our own, but we don't really have the resources for that so best to stick with something updated.

**Comment by phit** [comment]
what toml lib is forge gonna use? might be worth using the same one

**Comment by parlough** [comment]
They'll be using this https://github.com/TheElectronWill/Night-Config/tree/master/toml/src/main/java/com/electronwill/nightconfig/toml I believe, which takes a pretty similar position to Configurate in its features, so I'm not sure it's a good idea to include. Plus if we stay separate we can avoid dependency conflict issues like we had with the hocon library in the past.

**Comment by lucko** [comment]
cava is the best implementation I've seen so far.

Will give it a try at some point tomorrow, or maybe later today.

**Comment by kashike** [comment]
A TOML library that supports comments is pretty much a requirement, IMO.

File: configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLConfigurationLoader.java (hunk 3 above)
Comment (randombyte-developer) [reply configurate-toml/src/main/java/ninja/leaping/configurate/toml/TOMLConfigurationLoader.java]: Kotlin guy spotted

**Comment by pie-flavor** [comment]
Update: Latest cava-toml still fails its own tests. It does not correctly remove leading newlines from triple-quoted strings or handle nested arrays. You can clone it and see for yourself.

**Comment by lucko** [comment]
Indeed, and as far as I can tell it's also still missing comment handling functionality.

File: configurate-toml/src/test/resources/example.toml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+[database]
+server = "192.168.1.1"
+ports = [ 8001, 8001, 8002 ]
+connection_max = 5000
+enabled = true
+
+[servers]
+
+  # You can indent as you please. Tabs or spaces. TOML don't care.
```
Comment (hugmanrique) [inline configurate-toml/src/test/resources/example.toml]: Minor nit
```suggestion
  # You can indent as you please. Tabs or spaces. TOML doesn't care.
```

**Comment by pie-flavor** [comment]
@hugmanrique ... this is the example file straight from https://github.con/toml-lang/toml

**Comment by pie-flavor** [comment]
@lucko what progress?

**Comment by kashike** [comment]
None, and this won't be merged until a TOML library that supports loading and saving with comments comes around.

**Comment by pie-flavor** [comment]
Apparently https://github.com/TheElectronWill/Night-Config 's TOML module does this, but I don't know if it's a good idea to pull in another general config framework to complete our general config framework.

**Comment by pie-flavor** [comment]
Cava is no longer broken, I was able to get them to resolve the issue.

**Comment by pie-flavor** [comment]
Dunno why it closed there.

**Comment by pie-flavor** [comment]
@kashike we now have a toml library that supports loading and saving with comments

**Comment by felixoi** [comment]
IMO not optimal to base on a library of night-config since we loose a bit of independence but it's the best opinion for now I guess.

**Comment by pie-flavor** [comment]
Independence from what? We weren't exactly going to roll our own parser lol @felixoi

**Comment by felixoi** [comment]
Yeah I see your point but it's different IMO here since we depend on Night-Config. And Night-Config is a library pretty comparable to configurate if I don't miss something. So we kinda depend on another full config solution but we try to build an own one here.

**Comment by ryantheleach** [comment]
@felixoi Honestly that's all Configurates ever been from my perspective. An opinionated abstraction layer over specific config library implementations.

**Comment by zml2008** [comment]
I'm afraid at this point toml is not going to make it into configurate proper with any of the libraries currently available, and this PR has not seen progress in quite a while.

I still like the idea of TOML compatibility, and would happily merge a PR that provided that implementation using a library that could round-trip comments.

**Pull request closed without merging**